From 62539774106e551c6d547372ed06e7781bc67b8c Mon Sep 17 00:00:00 2001
From: Jiri Slaby <jirislaby@gmail.com>
Date: Mon, 8 Oct 2018 16:42:42 +0200
Subject: [PATCH] psicon: reconnect even if the original QNetworkSession is
 gone

When you suspend on LAN and resume on WLAN, psi+ won't reconnect as
QNetworkSession (corresponding to LAN) is closed externally.

When this happens, handle the close such that we open a new
QNetworkSession using the default configuration at that moment. And
when that becomes connected, psi+ reconnects.

This is mostly a work-around, merely a hack. We need to implement
roaming (or something alike) in the longterm.

This fixes #736.

--- git.orig/src/psicon.cpp
+++ git/src/psicon.cpp
@@ -385,9 +385,7 @@ bool PsiCon::init()
     PGPUtil::instance();
 #endif
 
-    d->netSession = new QNetworkSession(d->netConfMng.defaultConfiguration(), this);
-    connect(d->netSession, &QNetworkSession::opened, this, &PsiCon::networkSessionOpened);
-    d->netSession->open();
+    initNetSession();
 
     d->contactList = new PsiContactList(this);
 
@@ -1889,6 +1887,28 @@ void PsiCon::doWakeup()
     d->wakeupPending = true; // and wait for signal till network session is opened (proved to work on gentoo+nm+xfce)
 }
 
+void PsiCon::networkSessionClosed()
+{
+    /*
+     * The original connection is gone. (E.g. suspend on LAN, wake with WLAN
+     * only.)
+     *
+     * This is mostly a work-around, merely a hack. We need to implement
+     * roaming (or something alike) in the longterm.
+     */
+    disconnect(d->netSession);
+    d->netSession->deleteLater();
+    initNetSession();
+}
+
+void PsiCon::initNetSession()
+{
+    d->netSession = new QNetworkSession(d->netConfMng.defaultConfiguration(), this);
+    connect(d->netSession, &QNetworkSession::opened, this, &PsiCon::networkSessionOpened);
+    connect(d->netSession, &QNetworkSession::closed, this, &PsiCon::networkSessionClosed);
+    d->netSession->open();
+}
+
 void PsiCon::networkSessionOpened()
 {
     if (d->wakeupPending) {
--- git.orig/src/psicon.h
+++ git/src/psicon.h
@@ -187,6 +187,7 @@ private slots:
     void secondsIdle(int);    
     void proceedWithSleep();
     void networkSessionOpened();
+    void networkSessionClosed();
 private:
     class Private;
     Private *d;
@@ -202,6 +203,7 @@ private:
     void promptUserToCreateAccount();
     QString optionsFile() const;
     void doQuit(int);
+    void initNetSession();
 };
 
 #endif
